graphbolt/src/csc_sampling_graph.cc

```diff
@@ -131,16 +131,150 @@ c10::intrusive_ptr<SampledSubgraph> CSCSamplingGraph::InSubgraph(
           : torch::nullopt);
 }
 
+NumPickFn GetNumPickFn(
+    int64_t fanout, bool replace,
+    const torch::optional<torch::Tensor>& probs_or_mask) {
+  if (!probs_or_mask.has_value()) {
+    return [fanout, replace](int64_t offset, int64_t num_neighbors) {
+      int64_t max_count = fanout == -1 ? num_neighbors : fanout;
+      if (replace) {
+        return torch::tensor({num_neighbors == 0 ? 0 : max_count});
+      } else {
+        return torch::tensor({std::min(num_neighbors, max_count)});
+      }
+    };
+  } else {
+    return [fanout, replace, probs_or_mask](
+               int64_t offset, int64_t num_neighbors) {
+      int64_t max_count = fanout == -1 ? num_neighbors : fanout;
+      int64_t nonzero_count = 0;
+      nonzero_count = torch::nonzero(probs_or_mask.value().slice(
+                                         0, offset, offset + num_neighbors))
+                          .size(0);
+      if (replace) {
+        return torch::tensor({nonzero_count == 0 ? 0 : max_count});
+      } else {
+        return torch::tensor({std::min(nonzero_count, max_count)});
+      }
+    };
+  }
+}
+
+NumPickFn GetNumPickFnByEtype(
+    const std::vector<int64_t>& fanouts, bool replace,
+    const torch::Tensor& type_per_edge,
+    const torch::optional<torch::Tensor>& probs_or_mask) {
+  return [replace, fanouts, probs_or_mask, &type_per_edge](
+             int64_t offset, int64_t num_neighbors) {
+    int64_t etype_begin = offset;
+    int64_t etype_end = offset;
+    const int64_t end = offset + num_neighbors;
+    torch::Tensor count_by_etype = torch::zeros(fanouts.size());
+    AT_DISPATCH_ALL_TYPES(
+        type_per_edge.scalar_type(), "NumPickFnByEtype", ([&] {
+          const scalar_t* type_per_edge_data =
+              type_per_edge.data_ptr<scalar_t>();
+          while (etype_begin < end) {
+            scalar_t etype = type_per_edge_data[etype_begin];
+            TORCH_CHECK(
+                etype >= 0 && etype < (int64_t)fanouts.size(),
+                "Etype values exceed the number of fanouts.");
+            auto etype_end_it = std::upper_bound(
+                type_per_edge_data + etype_begin, type_per_edge_data + end,
+                etype);
+            etype_end = etype_end_it - type_per_edge_data;
+            // Do sampling for one etype.
+            count_by_etype[etype] =
+                GetNumPickFn(fanouts[etype], replace, probs_or_mask)(
+                    etype_begin, etype_end - etype_begin)[0];
+            etype_begin = etype_end;
+          }
+        }));
+    return count_by_etype;
```

Comment:
For simplicity,  here just return a total count. Because when return by etype, data should be filed in a per-allocated memory, which is too complex for now.